diff --git a/kernel-open/Kbuild b/kernel-open/Kbuild
index d231a29b..c25f7a34 100644
--- a/kernel-open/Kbuild
+++ b/kernel-open/Kbuild
@@ -105,6 +105,10 @@ $(foreach _module, $(NV_KERNEL_MODULES), \
 ccflags-y += -I$(src)/common/inc
 ccflags-y += -I$(src)
 ccflags-y += -Wall $(DEFINES) $(INCLUDES) -Wno-cast-qual -Wno-format-extra-args
+ccflags-y += -fsanitize=kernel-address -fasan-shadow-offset=0xdffffc0000000000 --param asan-globals=1 --param asan-instrumentation-with-call-threshold=10000 --param asan-instrument-allocas=1 --param asan-stack=1
+# ccflags-y += -fsanitize=bounds-strict -fsanitize=shift -fsanitize=bool -fsanitize=enum
+ccflags-y += -fsanitize-coverage=trace-pc -fsanitize-coverage=trace-cmp
+ccflags-y += -Wall $(DEFINES) $(INCLUDES) -Wno-cast-qual -Wno-format-extra-args
 ccflags-y += -D__KERNEL__ -DMODULE -DNVRM
 ccflags-y += -DNV_VERSION_STRING=\"615.71.09\"
 
diff --git a/kernel-open/nvidia-uvm/uvm_common.c b/kernel-open/nvidia-uvm/uvm_common.c
index 5f842004..66f789bd 100644
--- a/kernel-open/nvidia-uvm/uvm_common.c
+++ b/kernel-open/nvidia-uvm/uvm_common.c
@@ -244,6 +244,9 @@ void on_uvm_assert(void)
 #ifdef __COVERITY__
     __coverity_panic__()
 #endif
+#ifdef PANIC_ON_ASSERT
+      panic("Panic on UVM assertion failure")
+#endif
 }
 
 NV_STATUS uvm_spin_loop(uvm_spin_loop_t *spin)
diff --git a/src/common/unix/common/inc/nv_assert.h b/src/common/unix/common/inc/nv_assert.h
index 8c62ef52..b97ed79f 100644
--- a/src/common/unix/common/inc/nv_assert.h
+++ b/src/common/unix/common/inc/nv_assert.h
@@ -54,6 +54,10 @@ extern "C" {
 
   void nvDebugAssert(const char *expString, const char *filenameString,
                      const char *funcString, const unsigned int lineNumber);
+#ifdef PANIC_ON_ASSERT
+  void panic(const char *fmt, ...);
+#endif
+  void dump_stack(void);
 
   /*
    * Assert that (exp) is TRUE.  We use 'if (exp) { } else { fail }'
@@ -61,6 +65,17 @@ extern "C" {
    * accidentally write nvAssert(foo = bar) instead of nvAssert(foo ==
    * bar).
    */
+#ifdef PANIC_ON_ASSERT
+  #define nvAssert(exp)                                            \
+    do {                                                           \
+        if (exp) {                                                 \
+        } else {                                                   \
+            nvDebugAssert(#exp, __FILE__, __FUNCTION__, __LINE__); \
+            dump_stack();                                          \
+            panic("nvAssert failed");                              \
+        }                                                          \
+    } while (0)
+#else
   #define nvAssert(exp)                                            \
     do {                                                           \
         if (exp) {                                                 \
@@ -68,6 +83,7 @@ extern "C" {
             nvDebugAssert(#exp, __FILE__, __FUNCTION__, __LINE__); \
         }                                                          \
     } while (0)
+#endif
 
 #else
 
diff --git a/src/nvidia-modeset/Makefile b/src/nvidia-modeset/Makefile
index 9f7de4be..a833aa0f 100644
--- a/src/nvidia-modeset/Makefile
+++ b/src/nvidia-modeset/Makefile
@@ -129,6 +129,9 @@ CFLAGS += -fomit-frame-pointer
 CFLAGS += -fno-strict-aliasing
 CFLAGS += -ffreestanding
 CFLAGS += -fno-stack-protector
+CFLAGS += -fsanitize=kernel-address -fasan-shadow-offset=0xdffffc0000000000 --param asan-globals=1 --param asan-instrumentation-with-call-threshold=10000 --param asan-instrument-allocas=1 --param asan-stack=1
+# CFLAGS += -fsanitize=bounds-strict -fsanitize=shift -fsanitize=bool -fsanitize=enum
+CFLAGS += -fsanitize-coverage=trace-pc -fsanitize-coverage=trace-cmp
 
 CONDITIONAL_CFLAGS += $(call TEST_CC_ARG, -Wformat-overflow=2)
 CONDITIONAL_CFLAGS += $(call TEST_CC_ARG, -Wformat-truncation=1)
diff --git a/src/nvidia/Makefile b/src/nvidia/Makefile
index 9abfc7cf..1b7fee73 100644
--- a/src/nvidia/Makefile
+++ b/src/nvidia/Makefile
@@ -75,6 +75,9 @@ CFLAGS += -Wundef
 CFLAGS += -fno-common
 CFLAGS += -ffreestanding
 CFLAGS += -fno-stack-protector
+CFLAGS += -fsanitize=kernel-address -fasan-shadow-offset=0xdffffc0000000000 --param asan-globals=1 --param asan-instrumentation-with-call-threshold=10000 --param asan-instrument-allocas=1 --param asan-stack=1
+# CFLAGS += -fsanitize=bounds-strict -fsanitize=shift -fsanitize=bool -fsanitize=enum
+CFLAGS += -fsanitize-coverage=trace-pc -fsanitize-coverage=trace-cmp
 
 ifeq ($(TARGET_ARCH),x86_64)
   CFLAGS += -msoft-float
diff --git a/src/nvidia/inc/libraries/utils/nvassert.h b/src/nvidia/inc/libraries/utils/nvassert.h
index b1a8b2d8..7630887b 100644
--- a/src/nvidia/inc/libraries/utils/nvassert.h
+++ b/src/nvidia/inc/libraries/utils/nvassert.h
@@ -119,6 +119,10 @@ void __coverity_panic__(void);
 #define COVERITY_ASSERT_FAIL() __coverity_panic__()
 #else  // defined(__COVERITY__)
 #define COVERITY_ASSERT_FAIL() ((void) 0)
+#ifdef PANIC_ON_ASSERT
+void panic(const char *fmt, ...);
+#endif
+void dump_stack(void);
 #endif // defined(__COVERITY__)
 #endif // !defined(COVERITY_ASSERT_FAIL)
 
@@ -158,14 +162,37 @@ void nvAssertDestroy(void);
 void nvAssertFailed(NV_ASSERT_FAILED_FUNC_TYPE);
 void nvAssertOkFailed(NvU32 status NV_ASSERT_FAILED_FUNC_COMMA_TYPE);
 
+#ifdef PANIC_ON_ASSERT
 #define NV_ASSERT_FAILED(exprStr)                                              \
     do {                                                                       \
         NV_LOG_SPECIAL(LEVEL_ERROR, RM_GSP_LOG_SPECIAL_ASSERT_FAILED,          \
                        exprStr "\n");                                          \
         nvAssertFailed(0);                                                     \
+        dump_stack();                                                          \
+        panic("NV_ASSERT_FAILED");                                             \
         PORT_BREAKPOINT();                                                     \
     } while(0)
+#else
+#define NV_ASSERT_FAILED(exprStr)                                              \
+    do {                                                                       \
+        NV_LOG_SPECIAL(LEVEL_ERROR, RM_GSP_LOG_SPECIAL_ASSERT_FAILED,          \
+                       exprStr "\n");                                          \
+        nvAssertFailed(0);                                                     \
+        PORT_BREAKPOINT();                                                     \
+    } while(0)
+#endif
 
+#ifdef PANIC_ON_ASSERT
+#define NV_ASSERT_OK_FAILED(exprStr, status)                                   \
+    do {                                                                       \
+        NV_LOG_SPECIAL(LEVEL_ERROR, RM_GSP_LOG_SPECIAL_ASSERT_OK_FAILED,       \
+                       exprStr "\n", status);                                  \
+        nvAssertOkFailed(status, 0);                                           \
+        dump_stack();                                                          \
+        panic("NV_ASSERT_OK_FAILED");                                          \
+        PORT_BREAKPOINT();                                                     \
+    } while(0)
+#else
 #define NV_ASSERT_OK_FAILED(exprStr, status)                                   \
     do {                                                                       \
         NV_LOG_SPECIAL(LEVEL_ERROR, RM_GSP_LOG_SPECIAL_ASSERT_OK_FAILED,       \
@@ -173,6 +200,7 @@ void nvAssertOkFailed(NvU32 status NV_ASSERT_FAILED_FUNC_COMMA_TYPE);
         nvAssertOkFailed(status, 0);                                           \
         PORT_BREAKPOINT();                                                     \
     } while(0)
+#endif
 
 #define NV_CHECK_FAILED(level, exprStr)                                        \
    do {                                                                        \
@@ -216,21 +244,45 @@ void nvCheckOkFailedNoLog(NvU32 level, NvU32 status NV_ASSERT_FAILED_FUNC_COMMA_
     NVLOG_PRINTF(NV_PRINTF_MODULE, NVLOG_ROUTE_RM, level,                      \
     NV_PRINTF_ADD_PREFIX(fmt), ##__VA_ARGS__)
 
+#ifdef PANIC_ON_ASSERT
+#define NV_ASSERT_FAILED(exprStr)                                              \
+    do {                                                                       \
+        NV_ASSERT_LOG(LEVEL_ERROR, "Assertion failed: " exprStr);              \
+        nvAssertFailedNoLog(NV_ASSERT_FAILED_FUNC_PARAM(exprStr));             \
+        dump_stack();                                                          \
+        panic("NV_ASSERT_FAILED");                                             \
+        PORT_BREAKPOINT_CHECKED();                                             \
+    } while(0)
+#else
 #define NV_ASSERT_FAILED(exprStr)                                              \
     do {                                                                       \
         NV_ASSERT_LOG(LEVEL_ERROR, "Assertion failed: " exprStr);              \
         nvAssertFailedNoLog(NV_ASSERT_FAILED_FUNC_PARAM(exprStr));             \
         PORT_BREAKPOINT_CHECKED();                                             \
     } while(0)
+#endif
 
+#ifdef PANIC_ON_ASSERT
 #define NV_ASSERT_OK_FAILED(exprStr, status)                                   \
     do {                                                                       \
         NV_ASSERT_LOG(LEVEL_ERROR, "Assertion failed: 0x%08X returned from "   \
             exprStr, status);                                                  \
         nvAssertOkFailedNoLog(status                                           \
             NV_ASSERT_FAILED_FUNC_COMMA_PARAM(exprStr));                       \
+        dump_stack();                                                          \
+        panic("NV_ASSERT_OK_FAILED");                                          \
         PORT_BREAKPOINT_CHECKED();                                             \
     } while(0)
+#else
+#define NV_ASSERT_OK_FAILED(exprStr, status)                                   \
+    do {                                                                       \
+        NV_ASSERT_LOG(LEVEL_ERROR, "Assertion failed: 0x%08X returned from "   \
+            exprStr, status);                                                  \
+        nvAssertOkFailedNoLog(status                                           \
+            NV_ASSERT_FAILED_FUNC_COMMA_PARAM(exprStr));                       \
+        PORT_BREAKPOINT_CHECKED();                                             \
+    } while(0)
+#endif
 
 #define NV_CHECK_FAILED(level, exprStr)                                        \
     do {                                                                       \
@@ -253,17 +305,37 @@ void nvCheckOkFailedNoLog(NvU32 level, NvU32 status NV_ASSERT_FAILED_FUNC_COMMA_
         }                                                                      \
     } while(0)
 
+#ifdef PANIC_ON_ASSERT
 #define NV_ASSERT_FAILED_FUNC(exprStr)                                         \
     do {                                                                       \
         nvAssertFailed(NV_ASSERT_FAILED_FUNC_PARAM(exprStr));                  \
+        dump_stack();                                                          \
+        panic("NV_ASSERT_FAILED_FUNC");                                        \
         PORT_BREAKPOINT_CHECKED();                                             \
     } while(0)
+#else
+#define NV_ASSERT_FAILED_FUNC(exprStr)                                         \
+    do {                                                                       \
+        nvAssertFailed(NV_ASSERT_FAILED_FUNC_PARAM(exprStr));                  \
+        PORT_BREAKPOINT_CHECKED();                                             \
+    } while(0)
+#endif
 
+#ifdef PANIC_ON_ASSERT
+#define NV_ASSERT_OK_FAILED_FUNC(exprStr, status)                              \
+    do {                                                                       \
+        nvAssertOkFail(status NV_ASSERT_FAILED_FUNC_COMMA_PARAM(exprStr));     \
+        dump_stack();                                                          \
+        panic("NV_ASSERT_OK_FAILED_FUNC");                                     \
+        PORT_BREAKPOINT_CHECKED();                                             \
+    } while(0)
+#else
 #define NV_ASSERT_OK_FAILED_FUNC(exprStr, status)                              \
     do {                                                                       \
         nvAssertOkFail(status NV_ASSERT_FAILED_FUNC_COMMA_PARAM(exprStr));     \
         PORT_BREAKPOINT_CHECKED();                                             \
     } while(0)
+#endif
 
 #define NV_CHECK_FAILED_FUNC(level, exprStr)                                   \
     if (NV_PRINTF_LEVEL_ENABLED(level))                                        \
